### DISCUSSION TITLE
Possibility to add filter to entire canvas same as for individual objects

### DISCUSSION THREAD

**Discussion by shashank-brightness** [fabricjs/fabric.js#9755] [Show and tell]
### CheckList

- [X] I agree to follow this project's Code of Conduct
- [X] I have read and followed the [Contributing Guide](https://github.com/fabricjs/fabric.js/blob/master/CONTRIBUTING.md)
- [X] I have searched and referenced existing issues, feature requests and discussions
- [X] I am filing a **FEATURE** request.

### Description

When we need to add filter to entire canvas, nothing is currently available to do this. I need the same thing how we add to objects but for entire canvas.

```
const activeObj = canvas.getActiveObject();
const objectBrightnessFilter = new fabric.Image.filters.Brightness({
    brightness: 0.5
});
activeObj.filters.push(objectBrightnessFilter)
activeObj.applyFilters();
canvas.renderAll();
```

Above code adds brightness filter but for object only. We may be iterate over all objects but that does not look the proper way to do it. Also, getting whole image first via `toDataUrl` and applying filter to that image is not viable solution.

### Current State

Searched but could not found any solution to this issue.

**Comment by ShaMan123** [comment]
Try using the after:render canvas hook to do that
You can use the image filter logic

**Comment by shashank-brightness** [comment]
Can you please share how image filter thing would be using after:render? i am not familiar with that.

**Comment by ShaMan123** [comment]
Image filters take an image source, canvas is an image source.
So you can use canvas .getElement() as if it were an image and filter it.

**Comment by shashank-brightness** [comment]
I am not getting your last comment and I searched but could not found the example so would you mind sharing the actual code please? I am looking for example using fabric js way in webgl.

**Comment by asturur** [comment]
To be clear your description is very confusing and is unclear what you want to achieve.
FabricJS does not support filters on objects different than images ( even if a pr for that exists and at some point it will be integrated ).

Is unclear if you want to apply that dynamically or just on some kind of export and what kind of performances do you expect.
For objects that mutate quickly uploading the pixels to the gpu is slow so performances aren't great.

The first thing you can do is clone your objects as image and filter then the images.

If you want to apply the filter to the canvas surface you have to do some kind of more elaborate hack involving a proxy fabricImage and an after:render event and some more stuff. Not going to explain that if i m not sure that is what you look for

**Comment by shashank-brightness** [comment]
@asturur 
i have some sliders and i need to adjust brightness/contrast upon users changing that slider value but i don't want to export whole image via some function like `toDataUrl()` and apply filter on that exported image. 

I have already read this thing but not sure if something better exists than this. 
https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/filter

I need filter for whole canvas. I can't better explain than this.

Also, let's say if object is something else(text) other than the image, it should be that only after applying filter and not image. So iterating over everything and adding them back as image and apply filter is too long and would not work as well.

Let me know if fabric js can change like that for whole canvas if something is available on fabric's native code.

**Comment by shashank-brightness** [comment]
@asturur 
I need something like this which applies filter to everything below it (even to the background itself).
![Screenshot 2024-03-21 151637](https://github.com/fabricjs/fabric.js/assets/141903260/6c041519-3e56-46ff-abcb-811eade2b2a3)

**Comment by asturur** [comment]
Provide me a jsfiddle or a codesandbox ( that works ) with a simple canvas and a rect and a text and i ll see if i can make an example

**Reply by shashank-brightness** [reply]
@asturur
I have created here. 
https://jsfiddle.net/30anyoqc/

**Comment by ShaMan123** [comment]
If applying a filter to all your canvas is good enough then this is enough
```js
      canvas.on('before:render', ({
        ctx
      }) => {
        ctx.filter = 'blur(5px)'
      })
```
https://jsfiddle.net/tbp7gzru/ 


If not you will need to use layering, that is possible but you will need to impl on top of fabric.
I thought you wanted to use fabric's image filters as @asturur explained

**Reply by shashank-brightness** [reply]
Thanks for the code, I will use like that, however i would be interested know how it would be to use layering on top of fabric. you mean apply effect on each individual layer and also to background or something like that i guess?

Also, if i need to add both brightness and contrast to server via canvas.toJSON, are these things saved auto in json?

**Reply by wangyong991** [reply]
I want to apply the following fragmentSource to all canvas to achieve more complex effects instead of using the effects in CanvasRenderingContext2D.filter. How should I do this?
`precision highp float;
        uniform sampler2D uTexture;
        varying vec2 vTexCoord;
        void main() {
          vec4 color = texture2D(uTexture, vTexCoord);
          color.r = 0.0;
          color.g = 1.0;
          color.b = 1.0;
          gl_FragColor = color;
        }`

**Reply by asturur** [reply]
We don't support filter outside the image yet. You could try to make the filter for the image, than add an overlay to the canvas that is an image that takes the element from the canvas lower element and see what happens.

Make the filter work on the image, then think of the rest.
There is a filter boilerplate to create new filters that you can use

**Comment by ShaMan123** [comment]
> Thanks for the code, I will use like that, however i would be interested know how it would be to use layering on top of fabric. you mean apply effect on each individual layer and also to background or something like that i guess?

yes, that requires you to rebuild rendering
might be easier to use a rendering engine such as pixi.js or rive

> Also, if i need to add both brightness and contrast to server via canvas.toJSON, are these things saved auto in json?

no
